--- a/com/mojang/math/OctahedralGroup.java
+++ b/com/mojang/math/OctahedralGroup.java
@@ -98,6 +_,13 @@
     private static final OctahedralGroup[] INVERSE_TABLE = Arrays.stream(values())
         .map(f -> Arrays.stream(values()).filter(s -> f.compose(s) == IDENTITY).findAny().get())
         .toArray(OctahedralGroup[]::new);
+    // Paper start - Avoid Lazy Initialization for Enum Fields
+    static {
+        for (OctahedralGroup octahedralGroup : values()) {
+            octahedralGroup.initializeRotationDirections();
+        }
+    }
+    // Paper end - Avoid Lazy Initialization for Enum Fields
 
     OctahedralGroup(String name, SymmetricGroup3 permutation, boolean invertX, boolean invertY, boolean invertZ) {
         this.name = name;
@@ -139,7 +_,7 @@
         return this.name;
     }
 
-    public Direction rotate(Direction direction) {
+    public void initializeRotationDirections() { // Paper - Avoid Lazy Initialization for Enum Fields
         if (this.rotatedDirections == null) {
             this.rotatedDirections = Util.makeEnumMap(Direction.class, facing -> {
                 Direction.Axis oldAxis = facing.getAxis();
@@ -149,7 +_,11 @@
                 return Direction.fromAxisAndDirection(newAxis, newDirection);
             });
         }
+    // Paper start - Avoid Lazy Initialization for Enum Fields
+    }
 
+    public Direction rotate(Direction direction) {
+        // Paper end - Avoid Lazy Initialization for Enum Fields
         return this.rotatedDirections.get(direction);
     }
 
