--- a/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
@@ -35,7 +_,9 @@
     ) {
         super(mob, mustSee, mustReach);
         this.targetType = targetType;
-        this.randomInterval = reducedTickDelay(randomInterval);
+        // Leaf start - configurable goal update interval
+        this.randomInterval = gr1mly4memes.slime.SlimeConfig.chanceTarget > 0 && randomInterval <= 10 ? reducedTickDelay(gr1mly4memes.slime.SlimeConfig.chanceTarget) : reducedTickDelay(randomInterval);
+        // Leaf end - configurable goal update interval
         this.setFlags(EnumSet.of(Goal.Flag.TARGET));
         this.targetConditions = TargetingConditions.forCombat().range(this.getFollowDistance()).selector(selector);
     }
@@ -72,7 +_,7 @@
 
     @Override
     public void start() {
-        this.mob.setTarget(this.target);
+        this.mob.setTarget(this.target, target instanceof ServerPlayer ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY); // CraftBukkit - reason
         super.start();
     }
 
