--- a/net/minecraft/client/renderer/block/MovingBlockRenderState.java
+++ b/net/minecraft/client/renderer/block/MovingBlockRenderState.java
@@ -22,6 +_,7 @@
     public @Nullable Holder<Biome> biome;
     public CardinalLighting cardinalLighting = CardinalLighting.DEFAULT;
     public LevelLightEngine lightEngine = LevelLightEngine.EMPTY;
+    public net.neoforged.neoforge.model.data.ModelData modelData = net.neoforged.neoforge.model.data.ModelData.EMPTY;
 
     @Override
     public CardinalLighting cardinalLighting() {
@@ -48,11 +_,23 @@
         return pos.equals(this.blockPos) ? this.blockState : Blocks.AIR.defaultBlockState();
     }
 
+    // Slime - add getBlockStateIfLoaded() override
+    @Override
+    public @Nullable BlockState getBlockStateIfLoaded(BlockPos pos) {
+        return this.getBlockState(pos);
+    }
+
     @Override
     public FluidState getFluidState(BlockPos pos) {
         return this.getBlockState(pos).getFluidState();
     }
 
+    // Slime - add getFluidIfLoaded() override
+    @Override
+    public @Nullable FluidState getFluidIfLoaded(BlockPos pos) {
+        return this.getFluidState(pos);
+    }
+
     @Override
     public int getHeight() {
         return 1;
@@ -61,5 +_,10 @@
     @Override
     public int getMinY() {
         return this.blockPos.getY();
+    }
+
+    @Override
+    public net.neoforged.neoforge.model.data.ModelData getModelData(BlockPos pos) {
+        return pos.equals(this.blockPos) ? this.modelData : net.neoforged.neoforge.model.data.ModelData.EMPTY;
     }
 }
